### PR TITLE
[Wasm GC] Fix precompute on GC data

### PR DESCRIPTION
**Pull request by kripken**
This fixes precomputation on GC after #3803 was too optimistic.

The issue is subtle. Precompute will repeatedly evaluate expressions and
propagate their values, flowing them around, and it ignores side effects
when doing so. For example:
```wat
(block
 ..side effect..
 (i32.const 1)
)
```
When we evaluate that we see there are side effects, but regardless of them
we know the value flowing out is 1. So we can propagate that value, if it is
assigned to a local and read elsewhere.

This is not valid for GC because `struct.new` and `array.new` have a "side
effect" that is noticeable in the result. Each time we call `struct.new` we get a
new struct with a new address, which `ref.eq` can distinguish. So when this
pass evaluates the same thing multiple times it will get a different result.

Also, we can't precompute a `struct.get` even if we know the struct, not unless
we know the reference has not escaped (where a call could modify it).

To avoid all that, do not precompute references, aside from the trivially safe ones
like nulls and function references (simple constants that are the same each time
we evaluate the expression emitting them).

`precomputeExpression()` had a minor bug which this fixes. It checked the type
of the expression to see if we can create a constant for it, but really it should
check the value - since (separate from this PR) we have no way to emit a
"constant" for a struct etc. Also that only matters if `replaceExpression` is true, that
is, if we are replacing with a constant; if we just want the value internally, we have
no limit on that.

Also add Literal support for comparing GC refs, which is used by `ref.eq`. Without
that tiny fix the tests here crash.

This adds a bunch of tests, many for corner cases that we don't handle (since
the PR makes us not propagate GC references). But they should be helpful
if/when we do, to avoid the mistakes in #3803

### CODE REVIEW

File: test/lit/passes/precompute-gc.wast [hunk 1]
```diff
@@ -34,4 +39,356 @@
...
+ ;; CHECK-NEXT:     (rtt.canon $empty)
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:    (struct.new_default_with_rtt $empty
+ ;; CHECK-NEXT:     (rtt.canon $empty)
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (local.get $tempresult)
+ ;; CHECK-NEXT: )
+ (func $propagate-unequal (result i32)
```
Comment (tlively) [inline test/lit/passes/precompute-gc.wast:375]: Is this the case that the previous PR did not correctly handle?

File: src/passes/Precompute.cpp [hunk 2]
```diff
@@ -226,16 +235,20 @@ struct Precompute
...
-      return Flow(NONCONSTANT_FLOW);
-    }
+    Flow flow;
     try {
-      return PrecomputingExpressionRunner(
-               getModule(), getValues, replaceExpression)
-        .visit(curr);
+      flow =
+        PrecomputingExpressionRunner(getModule(), getValues, replaceExpression)
+          .visit(curr);
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/Precompute.cpp:242]: It might return a breaking flow in general (nonconstant or due to something else). In that case it should not matter much what we return, but you're right that this is confusing, I'll fix it.

File: src/passes/Precompute.cpp [hunk 3]
```diff
@@ -286,6 +299,16 @@ struct Precompute
         if (setValues[set].isConcrete()) {
           continue; // already known constant
         }
+        // Precompute the value. Note that this executes the code from scratch
+        // each time we reach this point, and so we need to be careful about
+        // repeating side effects if those side effects are expressed *in the
+        // value*. A case where that can happen is GC data (each struct.new
+        // creates a new, unique struct, even if the data is equal), and so GC
+        // refs are disallowed in canEmitConstantFor().
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/Precompute.cpp:307]: Sure, will add.

File: test/lit/passes/precompute-gc.wast (hunk 1 above)
Comment (kripken) [reply test/lit/passes/precompute-gc.wast:375]: Yes, exactly.

**Review by tlively — APPROVED** [review]
